src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Queries/SearchProductQueryHandler.cs

```diff
@@ -1,85 +1,259 @@
+                        CategoryId = x.CategoryId,
+                        Quantity = 1,
+                        Code = x.Code,
+                        InStockQuantity = 0, //TODO: and this
+                        Outline = x.Outline
+                    }).ToList()
+                })
+                .ToList();
+
+            var promotionResults = await cartAggregate.EvaluatePromotionsAsync(context);
+            var promoRewards = promotionResults.Rewards.OfType<CatalogItemAmountReward>().ToArray();
+
+            var language = new Language(cultureName);
+
+            foreach (var expProduct in expProducts)
+            {
+                var productId = expProduct.Id;
+
+                var prices = expProduct.Prices;
+                var productPrices = prices
+                    .Where(price => price.ProductId == productId)
+                    // Apply Prices
+                    .Select(price =>
+                    {
+                        var currency = new Currency(language, price.Currency);
+
+                        return new ProductPrice(currency)
+                        {
+                            ProductId = productId,
+                            PricelistId = price.PricelistId,
+                            Currency = currency,
+                            ListPrice = new Money(price.List, currency),
+                            SalePrice = price.Sale == null
+                                 ? new Money(price.List, currency)
+                                 : new Money((decimal)price.Sale, currency),
+                            MinQuantity = price.MinQuantity
+                        };
+                    })
+                    .GroupBy(x => x.Currency)
+                    .Where(x => x.Any())
+                    // Apply TierPrices
+                    .Select(currencyGroup =>
+                    {
+                        var orderedPrices = currencyGroup
+                            .OrderBy(x => x.MinQuantity ?? 0)
+                            .ThenBy(x => x.ListPrice);
+
+                        var nominalPrice = orderedPrices.First();
+
+                        var tierPrices = orderedPrices.Select(x => new TierPrice(x.SalePrice, x.MinQuantity ?? 1));
+
+                        nominalPrice.TierPrices.AddRange(tierPrices);
+
+                        return nominalPrice;
+                    })
+                    .ToList();
+
+                productPrices.ApplyRewards(promoRewards);
+
+                expProduct.ProductPrices = productPrices;
+            }
+        }
+
+        private string[] BuildFields(IReadOnlyCollection<string> includeFields, out bool loadPrices)
+        {
+            var result = new List<string>();
+
+            // Add filds for __object
+            result.AddRange(includeFields.Concat(new[] { "id" }).Select(x => "__object." + x));
+
+            loadPrices = includeFields.Any(x => x.Contains("prices", StringComparison.OrdinalIgnoreCase));
```

Comment:
Probably better to extract this method into separate NeedLoadPrices, call it in Handle methods and pass its result as an ordinary param (not out) to BuildFields.

Reply:
This code needs refactor to create "context" with this fields and use separate service to implement build logic, because we can use different search engines. I added todo with description with this problem.